extensions/amp-user-notification/0.1/amp-user-notification.js

```diff
@@ -0,0 +1,309 @@
+      this.activated_ = true;
+    }
+    return Promise.resolve();
+  }
+
+  /**
+   * Executes a `GET` request to the href given on the `data-show-if-href`
+   * attribute.
+   * @param {string} cid
+   * @return {!Promise}
+   * @private
+   */
+  fetchShowEndpoint_(cid) {
+    /** @type {GetRequestMetadata} */
+    let getReq = {
+      body: {
+        elementId: this.elementId_,
+        externalCID: cid
+      }
+    };
+
+    return xhrFor(this.win_).fetchJson(this.showIfHref_, getReq);
+  }
+
+  /**
+   * Creates an POST to the specified `data-dismiss-href` url.
+   * @private
+   * @return {!Promise}
+   */
+  postDismissEnpoint_() {
+    return xhrFor(this.win_).fetchJson(this.dismissHref_, {
+      method: 'POST',
+      body: {
+        elementId: this.elementId
+      }
+    });
+  }
+
+  /**
+   * Success handler for `fetchShowEndpoint_`
+   * @param {GetResponseMetadata}
+   * @private
+   */
+  onFetchShowEndpointSuccess_(data) {
+    assert(typeof data.show_notification == 'boolean', '%s should ' +
+        'be a boolean', data.show_notification);
+    if (data.show_notification) {
+      this.show();
+    }
+    return data;
+  }
+
+  /**
+   * Generic rejection handler for amp-user-notification.
+   * @private
+   */
+  onRejection_(msg) {
+    console/*OK*/.error('Rejection ocurred on `amp-user-notification` ' +
+        `with html element id "${this.elementId_}". message: ${msg}`);
+  }
+
+  /**
+   * Get async cid service.
+   * @private
+   */
+  getAsyncCid_() {
+    return cidFor(this.win_).then((cid) => {
+      return cid.get('amp-user-notification',
+          Promise.resolve(), this.dialogPromise_);
+    });
+  }
+
+  /**
+   * Shows the current user notification. Adds the `.amp-active` class to
+   * the element.
+   * @export
+   */
+  show() {
+    this.element.classList.add('amp-active');
+  }
+
+  /**
+   * Hides the current user notification and invokes the `dialogResolve_`
+   * method. Removes the `.amp-active` class from the element.
+   * @export
+   */
+  dismiss() {
+    this.vsync_.mutate(() => {
+      this.element.classList.remove('amp-active');
+      this.element.classList.add('amp-hidden');
+      this.dialogResolve_();
+    });
+  }
+}
+
+
+/**
+ * UserNotification service that is a per window singleton on installation.
+ */
+class UserNotification {
+
+  constructor(window) {
+    this.win_ = window;
+
+    /** @private @const {!Object<UserNotificationDefer>} */
+    this.deferRegistry_ = Object.create(null);
+
+    this.glasspane = this.win_.document.createElement('div');
+    this.glasspane.classList.add('-amp-glasspane');
+    this.win_.document.body.appendChild(this.glasspane);
+  }
+
+  /**
+   * Retrieve a promise associated to an `amp-user-notification` component
+   * that is resolved when user agrees to the terms.
+   * @return {!Promise|null}
+   */
+  get(id) {
+    let defer = this.getElementDeferById_(id);
+    if (defer) {
+      return defer.promise;
+    }
+    return null;
```

Comment:
so now i always return a promise (since the registration vs the service request is racey), it will never resolve if an `amp-user-notification` instance never resolves it.

I did add an assert that an element with the `id` should exist in the dom.